src/backend/columnar/columnar_storage.c

```diff
@@ -0,0 +1,579 @@
+	PageHeader phdr = (PageHeader) page;
+
+	if (phdr->pd_upper < offset + len)
+	{
+		elog(ERROR,
+			 "attempt to read columnar data of length %d from offset %d of block %d of relation %d",
+			 len, offset, blockno, rel->rd_id);
+	}
+
+	memcpy_s(buf, phdr->pd_upper - offset, page + offset, len);
+	ReleaseBuffer(buffer);
+}
+
+
+/*
+ * WriteToBlock - append data to a block, initializing if necessary, and emit
+ * WAL. If 'clear' is true, always clear the data on the page and reinitialize
+ * it first, and offset must be SizeOfPageHeaderData. Otherwise, offset must
+ * be equal to pd_lower and pd_lower will be set to the end of the written
+ * data.
+ */
+static void
+WriteToBlock(Relation rel, BlockNumber blockno, uint32 offset, char *buf,
+			 uint32 len, bool clear)
+{
+	Buffer buffer = ReadBuffer(rel, blockno);
+	LockBuffer(buffer, BUFFER_LOCK_EXCLUSIVE);
+
+	Page page = BufferGetPage(buffer);
+	PageHeader phdr = (PageHeader) page;
+	if (PageIsNew(page) || clear)
+	{
+		PageInit(page, BLCKSZ, 0);
+	}
+
+	if (phdr->pd_lower != offset || phdr->pd_upper - offset < len)
+	{
+		elog(ERROR,
+			 "attempt to write columnar data of length %d to offset %d of block %d of relation %d",
+			 len, offset, blockno, rel->rd_id);
+	}
+
+	START_CRIT_SECTION();
+
+	memcpy_s(page + phdr->pd_lower, phdr->pd_upper - phdr->pd_lower, buf, len);
+	phdr->pd_lower += len;
+
+	MarkBufferDirty(buffer);
+
+	if (RelationNeedsWAL(rel))
+	{
+		XLogBeginInsert();
+
+		/*
+		 * Since columnar will mostly write whole pages we force the transmission of the
+		 * whole image in the buffer
+		 */
+		XLogRegisterBuffer(0, buffer, REGBUF_FORCE_IMAGE);
+
+		XLogRecPtr recptr = XLogInsert(RM_GENERIC_ID, 0);
+		PageSetLSN(page, recptr);
+	}
+
+	END_CRIT_SECTION();
+
+	UnlockReleaseBuffer(buffer);
+}
+
+
+/*
+ * AlignReservation - given an unused logical byte offset, align it so that it
+ * falls at the start of a page.
+ *
+ * XXX: Reconsider whether we want/need to do this at all.
+ */
+static uint64
+AlignReservation(uint64 prevReservation)
+{
+	SmgrAddr prevAddr = logical_to_smgr(prevReservation);
+	uint64 alignedReservation = prevReservation;
+
+	if (prevAddr.offset != SizeOfPageHeaderData)
+	{
+		/* not aligned; align on beginning of next page */
+		SmgrAddr initial = { 0 };
+		initial.blockno = prevAddr.blockno + 1;
+		initial.offset = SizeOfPageHeaderData;
+		alignedReservation = smgr_to_logical(initial);
+	}
+
+	Assert(alignedReservation >= prevReservation);
+	return alignedReservation;
+}
+
+
+/*
+ * ColumnarMetapageIsCurrent - is the metapage at the latest version?
+ */
+static bool
+ColumnarMetapageIsCurrent(ColumnarMetapage *metapage)
+{
+	return (metapage->versionMajor == COLUMNAR_VERSION_MAJOR &&
+			metapage->versionMinor == COLUMNAR_VERSION_MINOR);
+}
+
+
+/*
+ * ColumnarMetapageCheckVersion - throw an error if accessing old
+ * version of metapage.
+ */
+static void
+ColumnarMetapageCheckVersion(Relation rel, ColumnarMetapage *metapage)
```

Comment:
Now we call this function for insert/truncate/vacuum codepaths.

However, there are some additional places that we need to consider, like: 
* Read path (maybe not needed now but will be when we implement tid mapping)
* Columnar udf's like `alter_columnar_table_set`

For the second one, maybe we can call `ColumnarMetapageCheckVersion` in `ReadColumnarOptions` too.
For read path, maybe we should call this either in `columnar_beginscan_extended` or `columnar_beginscan`.

But overall, there are a lot places that we should be careful about, so I'm still inclined to upgrade all columnar tables during upgrade but we can discuss that.

Reply:
Now this is called in the read path.

We could check it in other places, but that wouldn't seem to gain much. If there's a table with an old metapage, it's very likely to be caught by rudimentary tests. The problem before was that it wasn't caught in the read path, which is all a follower test can do.